### PR TITLE
WIP: Vertex tool improvement

### PR DESCRIPTION
**Pull request by mohsenD98**
### Description:
_coming soon..._

### Tasks Todo:   
- [ ] Logic based on finding a point not degree.  
- [ ] Move logic to utils class.  
- [ ] Better variable names.  
- [ ] More on documentations. 

### Maybe later
- Better visualization

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 31addf25ece1db6ade14d50cdac5426b6480406d: [**arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5797-5093b5f5a18c8809350735992bdf3d4128562d02-android-arm64.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5797-5093b5f5a18c8809350735992bdf3d4128562d02-android-arm-neon.apk)
- [x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5797-5093b5f5a18c8809350735992bdf3d4128562d02-android-x64.apk)
- [x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5797-5093b5f5a18c8809350735992bdf3d4128562d02-android-x86.apk)
- [**all access arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5797-5093b5f5a18c8809350735992bdf3d4128562d02-android-arm64.apk)
- [all access arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5797-5093b5f5a18c8809350735992bdf3d4128562d02-android-arm-neon.apk)
- [all access x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5797-5093b5f5a18c8809350735992bdf3d4128562d02-android-x64.apk)
- [all access x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5797-5093b5f5a18c8809350735992bdf3d4128562d02-android-x86.apk)
</details>

File: src/core/vertexmodel.h [hunk 1]
```diff
@@ -49,6 +49,9 @@ class QFIELD_CORE_EXPORT VertexModel : public QAbstractListModel
     Q_PROPERTY( int ringCount READ ringCount NOTIFY ringCountChanged )
     //! determines if the model has changes
     Q_PROPERTY( bool dirty READ dirty NOTIFY dirtyChanged )
+
+    Q_PROPERTY( bool angleFonud READ angleFonud NOTIFY angleFonudChanged )
```
Comment (nirvn) [inline src/core/vertexmodel.h:53]: As discussed, this should be an int property called snappedAngle.

Also, don't forget documentation.

File: src/core/vertexmodel.h [hunk 2]
```diff
@@ -70,6 +73,9 @@ class QFIELD_CORE_EXPORT VertexModel : public QAbstractListModel
      */
     Q_PROPERTY( int currentVertexIndex READ currentVertexIndex WRITE setCurrentVertexIndex NOTIFY currentVertexIndexChanged )
 
+    Q_PROPERTY( int snapToCommonAngleDegrees READ snapToCommonAngleDegrees WRITE setSnapToCommonAngleDegrees NOTIFY snapToCommonAngleDegreesChanged )
```
Comment (nirvn) [inline src/core/vertexmodel.h:76]: Documentation

File: src/core/vertexmodel.h [hunk 3]
```diff
@@ -223,6 +231,7 @@ class QFIELD_CORE_EXPORT VertexModel : public QAbstractListModel
 
     //! \copydoc dirty
     bool dirty() const;
+    bool angleFonud() const;
```
Comment (nirvn) [inline src/core/vertexmodel.h:234]: Documentation

File: src/core/vertexmodel.h [hunk 4]
```diff
@@ -265,6 +274,10 @@ class QFIELD_CORE_EXPORT VertexModel : public QAbstractListModel
 
     bool canUndo();
 
+    double calculateAngle( const QgsPoint &a, const QgsPoint &b, const QgsPoint &c );
```
Comment (nirvn) [inline src/core/vertexmodel.h:277]: If we still use this function, let's move it to GeometryUtils

File: src/core/vertexmodel.h [hunk 5]
```diff
@@ -265,6 +274,10 @@ class QFIELD_CORE_EXPORT VertexModel : public QAbstractListModel
...
+    int snapToCommonAngleDegrees() const;
+    void setSnapToCommonAngleDegrees( int snapToCommonAngleDegrees );
```
Comment (nirvn) [inline src/core/vertexmodel.h:279]: Documentaton

File: src/core/vertexmodel.h [hunk 6]
```diff
@@ -350,6 +369,7 @@ class QFIELD_CORE_EXPORT VertexModel : public QAbstractListModel
     bool mHistoryTraversing = false;
 
     friend class VertexModelTest;
+    int mSnapToCommonAngleDegrees;
```
Comment (nirvn) [inline src/core/vertexmodel.h:372]: Keep the friend class line at the very bottom of the class definition, and make sure you initialize your members (i.e. in this case int mSnapToCommonAngleDegrees = 0;

File: src/core/vertexmodel.h [hunk 7]
```diff
@@ -187,6 +193,8 @@ class QFIELD_CORE_EXPORT VertexModel : public QAbstractListModel
     //! next vertex or segment
     Q_INVOKABLE void next();
 
+    Q_INVOKABLE QgsPoint getPoint( int index );
```
Comment (nirvn) [inline src/core/vertexmodel.h:196]: Do we need to expose this to QML? Should it be a private function?

File: src/core/vertexmodel.cpp [hunk 8]
```diff
@@ -685,6 +691,31 @@ QgsPoint VertexModel::currentPoint() const
 {
   return mVertices.value( mCurrentIndex ).point;
 }
+// Function to calculate the angle between two vectors
```
Comment (nirvn) [inline src/core/vertexmodel.cpp:694]: FYI, we don't add comments above functions, we document in the .h[eader] :)

File: src/core/vertexmodel.cpp [hunk 9]
```diff
@@ -696,6 +727,25 @@ void VertexModel::setCurrentPoint( const QgsPoint &point )
...
+  {
+    startPoint = vertexCount() + startPoint;
+  }
+  int endPoint = mCurrentIndex + 2;
+  if ( endPoint >= vertexCount() )
+  {
+    endPoint = endPoint - vertexCount();
+  }
+
+  double angle = calculateAngle( mVertices[startPoint].point, mVertices[mCurrentIndex].point, mVertices[endPoint].point );
```
Comment (nirvn) [inline src/core/vertexmodel.cpp:743]: Alright, as discussed yesterday, here I think we shouldn't calculate the angle but rather the point location of the point that'd form the angle we're looking for. 

Then, we will snap to that point whenever we are close to it.

File: src/core/vertexmodel.cpp [hunk 10]
```diff
@@ -696,6 +727,25 @@ void VertexModel::setCurrentPoint( const QgsPoint &point )
...
+
+  qDebug() << "--------- angle = " << angle;
+
+  setAngleFonud( angle > snapToCommonAngleDegrees() - 1 && angle < snapToCommonAngleDegrees() + 1 );
```
Comment (nirvn) [inline src/core/vertexmodel.cpp:747]: We discussed this yesterday. This wouldn't really work as a/ we wouldn't get a perfect coordinate snapped to angle in any case, and b/ we'd allow for angle 89 and 91 :scream: :wink:

File: src/core/vertexmodel.cpp [hunk 11]
```diff
@@ -881,6 +936,15 @@ void VertexModel::setDirty( bool dirty )
   emit dirtyChanged();
 }
 
+void VertexModel::setAngleFonud( bool angleFonud )
```
Comment (nirvn) [inline src/core/vertexmodel.cpp:939]: This becomes a int commented int he header. But also, keep the getter and setter functions together in the source code (ATM they are disjointed and it just makes the code harder to skim through).

**Pull request closed without merging**